scrapy/dupefilters.py

```diff
@@ -38,10 +39,32 @@ def __init__(self, path=None, debug=False):
             self.file.seek(0)
             self.fingerprints.update(x.rstrip() for x in self.file)
 
+    @classmethod
+    def from_crawler(cls, crawler):
+        try:
+            dupefilter = cls.from_settings(crawler.settings)
+        except AttributeError:
+            debug = crawler.settings.getbool('DUPEFILTER_DEBUG')
+            fingerprinter = crawler.settings.getinstance(
+                'REQUEST_FINGERPRINTER',
+                crawler=crawler,
+                singleton=True,
+            )
+            dupefilter = cls(
+                job_dir(crawler.settings),
+                debug,
+                fingerprinter=fingerprinter,
+            )
+        return dupefilter
+
     @classmethod
     def from_settings(cls, settings):
         debug = settings.getbool('DUPEFILTER_DEBUG')
-        return cls(job_dir(settings), debug)
+        fingerprinter = settings.getinstance(
+            'REQUEST_FINGERPRINTER',
+            singleton=True,
+        )
+        return cls(job_dir(settings), debug, fingerprinter=fingerprinter)
```

Comment:
Is from_settings method required just for backwards compatibility, or is there another reason code is duplicated here?

Reply:
Backward compatibility (subclasses calling `super().from_settings()`?). I must say that it is not that clear to me how inheritance and these class methods should be handled.

If you are OK with it, I’ll just drop the class method. Otherwise, things get really messy.